### ISSUE TITLE
Telemetry for non-event features

### ISSUE THREAD

**Issue by MeltyBot** [meltano/squared#115]
Migrated from GitLab: https://gitlab.com/meltano/squared/-/issues/129

Originally created by @pnadolny13 on 2022-05-18 23:23:55

---

- lock files
- interactive config
- other?

Potentially this is solved by the new event structure if it includes a features list.

**Comment by pnadolny13** [comment]
Relates to discussion https://github.com/meltano/meltano/discussions/6036#discussioncomment-3129766

**Comment by WillDaSilva** [comment]
One place that would probably be pretty easy to instrument would be feature flags, i.e. if a feature flag is enabled, we can include that information in the project context.

**Comment by pnadolny13** [comment]
@tayloramurphy as I'm thinking about this more I don't know of a sustainable way to add these as part of the event. I think it will mostly be up to me on the warehouse side to parse out whether a feature is used. If thats the case then I just need to know that I have enough data for each feature to confidently tag it. Thinking of features we care about I grouped them by whether we have enough data today to tag events:

Not parseable:
- state backends
- lock file

Parseable:
- environments
- stream maps
- interactive config

Not sure:
- containers

If its only a few one off feature, i.e. state backends, then we can add an explicit attribute to the environment or cli context. Like `state_backend: 's3/sqlite/postgres/cloud/etc.'`. Or for lock files we can add them to the plugins context.

What do you think?

cc @WillDaSilva

**Comment by WillDaSilva** [comment]
@pnadolny13 When it comes to things we can know at startup, the two ways I see to do this are:
- Adding fields to one of the contexts (likely the environment context)
- Creating a startup event that includes a bunch of one-off information we'd like to know about an execution, but don't want to include with each telemetry event (as we would if we added it to the environment context)

It becomes more complicated when multiple services and such need to be started in Meltano before we can get the desired info, or when the info we're after depends on something happening during runtime. For that kind of thing I prefer using events, since contexts _should_ be immutable (though we have broken this rule once).

To avoid an explosion of event types, maybe we could make a generic "feature event". It schema could be structured as a list of objects which have a string field with a feature name, and a `oneOf` block with the subschema for that kind of feature.

**Comment by aaronsteers** [comment]
@pnadolny13  - I don't see a context for a prioritization on this so can I assume relatively low pri?

Also, this seems like it should be several issues focusing on the multiple disparate use cases, and not a single issue. If we can solve generically the multiple use cases, then certainly we'd like to do so, but I'm not sure the above has enough context to inform if we've solved each in a meaningful way.

For example, lock files treated as a feature would be a boolean, which would be `1` whether all plugins are locked or if only 2 out of 10 are locked. Also, many features would not and probably should not be parsed eagerly from the tree at project initialization time. 

Of all the use cases defined above, the only one that fits the 'feature flag' definition would be `env var strict mode`, since that necessarily must be read at a certain time during project initialization, and therefor would be available for context in the project events. But for that use case, the strict mode flag, there's no need for new events, since that can be handled with existing events and (maybe existing contexts?) - at most we'd add a new context to existing events.

**Comment by tayloramurphy** [comment]
@aaronsteers this is relatively low priority at the moment. We'll need this post-GA but I'm glad we're thinking about it a bit right now. 

@pnadolny13 one possible model of this is the "Service Ping" model that GitLab (https://docs.gitlab.com/ee/development/service_ping/#) had for self-managed installs. When we didn't have access to the database we would get these summary events that basicaly were a large dictionary of ints, bools, and strings about the overall usage of an installation. 

All that to say - we likely need to open issues on specific features to get usage data on them, starting with state backends and lock files.

**Comment by aaronsteers** [comment]
Closing. 

Replaced by:

- #7217
- #7218
